// sandstone/_internals/commands/implementations/Advancement.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdvancementCommand = void 0;
const Command_1 = require("../Command");
const decorators_1 = require("../decorators");
class AdvancementArguments extends Command_1.Command {
    constructor() {
        super(...arguments);
        /** Adds or removes all loaded advancements. */
        this.everything = () => { };
        /**
         * Adds or removes a single advancement or criterion.
         *
         * @param advancement Specifies a valid namespaced id of the advancement to target.
         *
         * @param criterion Specifies a valid criterion of the advancement to manipulate.
         * The command defaults to the entire advancement.
         * If specified, the command refers to merely the criterion and not the entire advancement.
         */
        this.only = (advancement, criterion) => { };
        /**
         * Adds or removes an advancement and all its children advancements.
         * Think of specifying everything from that advancement to the end.
         *
         * The exact order the operation is carried out in is:
         *
         *     specified advancement > child > child's child > ...
         *
         * When it operates on a child that branches, it iterates through all its children before continuing.
         *
         * @param advancement Specifies a valid namespaced id of the advancement to target.
         */
        this.from = (advancement) => { };
        /**
         * Specifies an advancement, and adds or removes all its parent advancements, and all its children advancements.
         * Think of specifying everything through the specified advancement, going both backwards and forwards.
         *
         * The exact order the operation is as if the command were executed with "until" specified, then with "from" specified:
         *
         *     parent > parent's parent > ... > root > specified advancement > child > child's child > ...
         *
         * @param advancement Specifies a valid namespaced id of the advancement to target.
         */
        this.through = (advancement) => { };
        /**
         * Adds or removes an advancement and all its parent advancements until the root for addition/removal.
         * Think of specifying everything from the start until that advancement.
         *
         * The exact order the operation is carried out in is:
         *
         *     parent > parent's parent > ... > root > specified advancement.
         *
         * @param advancement Specifies a valid namespaced id of the advancement to target.
         */
        this.until = (advancement) => { };
    }
}
__decorate([
    decorators_1.command('everything', { isRoot: false, executable: true }),
    __metadata("design:type", Object)
], AdvancementArguments.prototype, "everything", void 0);
__decorate([
    decorators_1.command('only', { isRoot: false, executable: true }),
    __metadata("design:type", Object)
], AdvancementArguments.prototype, "only", void 0);
__decorate([
    decorators_1.command('from', { isRoot: false, executable: true }),
    __metadata("design:type", Object)
], AdvancementArguments.prototype, "from", void 0);
__decorate([
    decorators_1.command('through', { isRoot: false, executable: true }),
    __metadata("design:type", Object)
], AdvancementArguments.prototype, "through", void 0);
__decorate([
    decorators_1.command('until', { isRoot: false, executable: true }),
    __metadata("design:type", Object)
], AdvancementArguments.prototype, "until", void 0);
/** Gives or takes an advancement from one or more players. */
class AdvancementCommand extends Command_1.Command {
    constructor() {
        super(...arguments);
        /**
         * Adds specified advancements.
         *
         * @param targets Specifies one player or more.
         */
        this.grant = (targets) => new AdvancementArguments(this.commandsRoot);
        /**
         * Removes specified advancements.
         *
         * @param targets Specifies one player or more.
         */
        this.revoke = (targets) => new AdvancementArguments(this.commandsRoot);
    }
}
__decorate([
    decorators_1.command(['advancement', 'grant'], { isRoot: true, hasSubcommands: true, executable: false }),
    __metadata("design:type", Object)
], AdvancementCommand.prototype, "grant", void 0);
__decorate([
    decorators_1.command(['advancement', 'revoke'], { isRoot: true, hasSubcommands: true, executable: false }),
    __metadata("design:type", Object)
], AdvancementCommand.prototype, "revoke", void 0);
exports.AdvancementCommand = AdvancementCommand;
