// sandstone/_internals/commands/implementations/Datapack.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatapackCommand = void 0;
const Command_1 = require("../Command");
const decorators_1 = require("../decorators");
class DatapackEnable extends Command_1.Command {
    constructor() {
        super(...arguments);
        /**
         * Load this pack before all others. It will have the lowest priority.
         */
        this.first = () => { };
        /**
         * Load this pack after all others. It will have the highest priority.
         */
        this.last = () => { };
        /**
         * Load this pack just before an existing pack.
         * It will have a lower priority than the existing one.
         *
         * @param name The name of the existing datapack.
         */
        this.before = (name) => { };
        /**
         * Load this pack just after an existing pack.
         * It will have a higher priority than the existing one.
         *
         * @param name The name of the existing datapack.
         */
        this.after = (name) => { };
    }
}
__decorate([
    decorators_1.command('first'),
    __metadata("design:type", Object)
], DatapackEnable.prototype, "first", void 0);
__decorate([
    decorators_1.command('last'),
    __metadata("design:type", Object)
], DatapackEnable.prototype, "last", void 0);
__decorate([
    decorators_1.command('before'),
    __metadata("design:type", Object)
], DatapackEnable.prototype, "before", void 0);
__decorate([
    decorators_1.command('after'),
    __metadata("design:type", Object)
], DatapackEnable.prototype, "after", void 0);
/** Controls the loading/unloading of data packs. */
class DatapackCommand extends Command_1.Command {
    constructor() {
        super(...arguments);
        /**
         * Disable the specified pack.
         *
         * @param name Specifies the name of the data pack.
         */
        this.disable = (name) => { };
        /**
         * Enable the specified pack.
         *
         * @param name Specifies the name of the data pack.
         *
         * @example
         *
         * // Enable `myDatapack`, and give it the highest priority
         * datapack.enable('myDatapack').last()
         *
         * // Enable `myDatapack`, just before `importantDatapack`.
         * // `mydatapack` will have a lower priority.
         * datapack.enable('myDatapack').before('importantDatapack')
         */
        this.enable = (name) => new DatapackEnable(this.commandsRoot);
        /** List all data packs, or list only the available/enabled ones.
         *
         * Hovering over the data packs in the chat output shows their description defined in their pack.mcmeta.
         *
         * @param typ `"available"` to only show available datapacks, `"enabled"` to only show enabled ones.
         */
        this.list = (type) => { };
    }
}
__decorate([
    decorators_1.command(['datapack', 'disable'], { isRoot: true }),
    __metadata("design:type", Object)
], DatapackCommand.prototype, "disable", void 0);
__decorate([
    decorators_1.command(['datapack', 'enable'], { isRoot: true, hasSubcommands: true }),
    __metadata("design:type", Object)
], DatapackCommand.prototype, "enable", void 0);
__decorate([
    decorators_1.command(['datapack', 'list'], { isRoot: true }),
    __metadata("design:type", Object)
], DatapackCommand.prototype, "list", void 0);
exports.DatapackCommand = DatapackCommand;
