// sandstone/_internals/variables/PlayerScore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerScore = void 0;
const abstractClasses_1 = require("./abstractClasses");
function createVariable(datapack, amountOrTargets, objective) {
    const anonymousScore = datapack.Variable();
    if (typeof amountOrTargets === 'number') {
        anonymousScore.set(amountOrTargets);
    }
    else {
        anonymousScore.set(amountOrTargets, objective);
    }
    return anonymousScore;
}
class PlayerScore extends abstractClasses_1.ComponentClass {
    constructor(commandsRoot, target, objective) {
        super();
        this.commandsRoot = commandsRoot;
        this.target = target;
        this.objective = objective;
    }
    toString() {
        return `${this.target} ${this.objective}`;
    }
    _toChatComponent() {
        return {
            score: { name: this.target, objective: this.objective.name },
        };
    }
    unaryOperation(operation, operator, ...args) {
        var _a;
        if (typeof args[0] === 'number') {
            this.commandsRoot.scoreboard.players[operation](this.target, this.objective, args[0]);
        }
        else if (args[0] instanceof PlayerScore) {
            this.commandsRoot.scoreboard.players.operation(this.target, this.objective, operator, args[0].target, args[0].objective);
        }
        else {
            this.commandsRoot.scoreboard.players.operation(this.target, this.objective, operator, args[0], (_a = args[1]) !== null && _a !== void 0 ? _a : this.objective);
        }
        return this;
    }
    binaryOperation(operator, ...args) {
        var _a;
        if (args[0] instanceof PlayerScore) {
            this.commandsRoot.scoreboard.players.operation(this.target, this.objective, operator, args[0].target, args[0].objective);
            return this;
        }
        if (typeof args[0] === 'number') {
            this.commandsRoot.Datapack.registerNewConstant(args[0]);
        }
        this.commandsRoot.scoreboard.players.operation(this.target, this.objective, operator, args[0], (_a = args[1]) !== null && _a !== void 0 ? _a : this.objective);
        return this;
    }
    set(...args) {
        return this.unaryOperation('set', '=', ...args);
    }
    add(...args) {
        return this.unaryOperation('add', '+=', ...args);
    }
    remove(...args) {
        return this.unaryOperation('remove', '-=', ...args);
    }
    multiply(...args) {
        return this.binaryOperation('*=', ...args);
    }
    divide(...args) {
        return this.binaryOperation('/=', ...args);
    }
    modulo(...args) {
        return this.binaryOperation('%=', ...args);
    }
    swap(...args) {
        this.binaryOperation('<>', ...args);
    }
    plus(...args) {
        const anonymousScore = createVariable(this.commandsRoot.Datapack, this.target, this.objective);
        anonymousScore.unaryOperation('add', '+=', ...args);
        return anonymousScore;
    }
    minus(...args) {
        const anonymousScore = createVariable(this.commandsRoot.Datapack, this.target, this.objective);
        anonymousScore.unaryOperation('remove', '-=', ...args);
        return anonymousScore;
    }
    multipliedBy(...args) {
        const anonymousScore = createVariable(this.commandsRoot.Datapack, this.target, this.objective);
        anonymousScore.binaryOperation('*=', ...args);
        return anonymousScore;
    }
    dividedBy(...args) {
        const anonymousScore = createVariable(this.commandsRoot.Datapack, this.target, this.objective);
        anonymousScore.binaryOperation('*=', ...args);
        return anonymousScore;
    }
    moduloBy(...args) {
        const anonymousScore = createVariable(this.commandsRoot.Datapack, this.target, this.objective);
        anonymousScore.binaryOperation('%=', ...args);
        return anonymousScore;
    }
    /** COMPARISONS OPERATORS */
    comparison(operator, matchesRange, args) {
        const playerScore = this;
        if (typeof args[0] === 'number') {
            return {
                _toMinecraftCondition: () => ({ value: ['score', playerScore.target, playerScore.objective, 'matches', matchesRange] }),
            };
        }
        const endArgs = args[1] ? args : [args[0]];
        return {
            _toMinecraftCondition: () => ({ value: ['score', playerScore.target, playerScore.objective, operator, ...endArgs] }),
        };
    }
    greaterThan(...args) {
        return this.comparison('>', `${typeof args[0] === 'number' ? args[0] + 1 : null}..`, args);
    }
    greaterOrEqualThan(...args) {
        return this.comparison('>=', `${args[0]}..`, args);
    }
    lowerThan(...args) {
        return this.comparison('<', `..${typeof args[0] === 'number' ? args[0] - 1 : null}`, args);
    }
    lowerOrEqualThan(...args) {
        return this.comparison('<=', `..${args[0]}`, args);
    }
    equalTo(...args) {
        return this.comparison('=', args[0].toString(), args);
    }
}
exports.PlayerScore = PlayerScore;
