// sandstone/_internals/resources/Advancement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Advancement = void 0;
const minecraft_1 = require("../datapack/minecraft");
class Advancement {
    constructor(commandsRoot, name, advancement) {
        this.commandsRoot = commandsRoot;
        this.advancementJson = advancement;
        this.path = this.commandsRoot.Datapack.getResourcePath(name);
        this.commandsRoot.Datapack.addResource(name, 'advancements', { advancement });
    }
    get name() {
        return minecraft_1.toMcFunctionName(this.path.fullPathWithNamespace);
    }
    /**
     * Grant this achievement to the players.
     */
    grant(players, criterion) {
        this.commandsRoot.advancement.grant(players).only(this.name, criterion);
    }
    /**
     * Revoke this achievement from the players.
     */
    revoke(players, criterion) {
        this.commandsRoot.advancement.revoke(players).only(this.name, criterion);
    }
    /**
     * Grant all achievements until this one to the players.
     */
    grantUntilThis(players) {
        this.commandsRoot.advancement.grant(players).until(this.name);
    }
    /**
     * Revoke all achievements until this one from the players.
     */
    revokeUntilThis(players) {
        this.commandsRoot.advancement.revoke(players).until(this.name);
    }
    /**
     * Grant all achievements from this one to the players.
     */
    grantFromThis(players) {
        this.commandsRoot.advancement.grant(players).from(this.name);
    }
    /**
     * Revoke all achievements from this one from the players.
     */
    revokeFromThis(players) {
        this.commandsRoot.advancement.revoke(players).from(this.name);
    }
}
exports.Advancement = Advancement;
