// sandstone/_internals/resources/LootTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LootTable = void 0;
const minecraft_1 = require("../datapack/minecraft");
class LootTable {
    constructor(datapack, name, lootTable) {
        this.give = (players) => this.datapack.commandsRoot.loot.give(players).loot(this.name);
        this.insert = (targetPos) => this.datapack.commandsRoot.loot.insert(targetPos).loot(this.name);
        this.replaceBlock = (targetPos, slot, count) => this.datapack.commandsRoot.loot.replaceBlock(targetPos, slot, count).loot(this.name);
        this.replaceEntity = (entities, slot, count) => this.datapack.commandsRoot.loot.replaceEntity(entities, slot, count).loot(this.name);
        this.lootTable = lootTable;
        this.path = datapack.getResourcePath(name);
        this.datapack = datapack;
        datapack.addResource(name, 'loot_tables', { lootTable });
    }
    get name() {
        return minecraft_1.toMcFunctionName(this.path.fullPathWithNamespace);
    }
}
exports.LootTable = LootTable;
