// sandstone/_internals/resources/Tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tag = void 0;
function isMcFunctionReturn(v) {
    return typeof v === 'function';
}
function isTagObject(v) {
    return typeof v === 'object';
}
class Tag {
    constructor(datapack, type, name, values, replace) {
        this.type = type;
        this.values = values.map((v) => {
            if (isMcFunctionReturn(v)) {
                return v.getName();
            }
            if (isTagObject(v) && isMcFunctionReturn(v.id)) {
                return {
                    id: v.id.getName(),
                    required: v.required,
                };
            }
            return v;
        });
        this.name = name;
        this.datapack = datapack;
        const { namespace, fullPath } = datapack.getResourcePath(name);
        datapack.resources.addResource('tags', {
            children: new Map(),
            isResource: true,
            path: [namespace, type, ...fullPath],
            values: this.values,
            replace,
        });
    }
}
exports.Tag = Tag;
